### PR TITLE
Make table mandatory for drop_chunks

### PR DESCRIPTION
**Pull request by mkindahl**
Make table mandatory for drop_chunks
    
    The `drop_chunks` function is refactored to make table name mandatory
    for the function. As a result, the function was also refactored to
    accept the `regclass` type instead of table name plus schema name and
    the parameters were reordered to match the order for `show_chunks`.
    
    The commit also refactor the code to pass the hypertable structure
    between internal functions rather than the hypertable relid and moving
    error checks to the PostgreSQL function.  This allow the internal
    functions to avoid some lookups and use the information in the
    structure directly and also give errors earlier instead of first
    dropping chunks and then error and roll back the transaction.

Fixes timescale/timescaledb-private#785 timescale/timescaledb-private#591

### CODE REVIEW

File: test/expected/alter.out [hunk 1]
```diff
@@ -303,8 +303,8 @@ DROP TABLESPACE tablespace1;
 ERROR:  tablespace "tablespace1" is still attached to 1 hypertables
 \set ON_ERROR_STOP 1
 -- show_chunks and drop_chunks output should be the same
-\set QUERY1 'SELECT show_chunks(\'hyper_in_space\', 22)::NAME'
-\set QUERY2 'SELECT drop_chunks(22, \'hyper_in_space\')::NAME'
+\set QUERY1 'SELECT show_chunks(\'hyper_in_space\', 22)::REGCLASS::TEXT'
```
Comment (erimatnor) [inline test/expected/alter.out:306]: Why cast TEXT to REGCLASS and then back to TEXT?

File: test/expected/alter.out [hunk 2]
```diff
@@ -303,8 +303,8 @@ DROP TABLESPACE tablespace1;
...
+\set QUERY2 'SELECT drop_chunks(\'hyper_in_space\', 22)::TEXT'
```
Comment (erimatnor) [inline test/expected/alter.out:307]: We're already returning TEXT, so this cast seems unnecessary.

File: test/sql/alter.sql [hunk 3]
```diff
@@ -167,8 +167,8 @@ DROP TABLESPACE tablespace1;
 \set ON_ERROR_STOP 1
 
 -- show_chunks and drop_chunks output should be the same
-\set QUERY1 'SELECT show_chunks(\'hyper_in_space\', 22)::NAME'
-\set QUERY2 'SELECT drop_chunks(22, \'hyper_in_space\')::NAME'
+\set QUERY1 'SELECT show_chunks(\'hyper_in_space\', 22)::REGCLASS::TEXT'
```
Comment (erimatnor) [inline test/sql/alter.sql:170]: Why cast TEXT to REGCLASS and then back to TEXT?

File: test/sql/alter.sql [hunk 4]
```diff
@@ -167,8 +167,8 @@ DROP TABLESPACE tablespace1;
...
+\set QUERY2 'SELECT drop_chunks(\'hyper_in_space\', 22)::TEXT'
```
Comment (erimatnor) [inline test/sql/alter.sql:171]: `drop_chunk` already returns TEXT so why is this case even needed?

File: test/sql/chunk_utils.sql [hunk 5]
```diff
@@ -171,8 +165,8 @@ SELECT * FROM show_chunks('drop_chunk_test2');
 
 \dt "_timescaledb_internal"._hyper*
 -- show_chunks and drop_chunks output should be the same
-\set QUERY1 'SELECT show_chunks(\'drop_chunk_test1\', older_than => 3)::NAME'
-\set QUERY2 'SELECT drop_chunks(3, \'drop_chunk_test1\')::NAME'
+\set QUERY1 'SELECT show_chunks(\'drop_chunk_test1\', older_than => 3)::REGCLASS::TEXT'
```
Comment (erimatnor) [inline test/sql/chunk_utils.sql:168]: More of this casting that seems unnecessary. I guess this applies also to a lot of places below.

File: tsl/test/sql/dist_hypertable.sql.in [hunk 6]
```diff
@@ -843,22 +843,22 @@ SELECT (_timescaledb_internal.show_chunk(show_chunks)).*
...
 
--- test passing newer_than as interval and cascade
-SELECT * FROM drop_chunks(newer_than => interval '10 years', table_name => 'disttable_drop_chunks', cascade => true);
+-- test passing newer_than as interval
+SELECT * FROM drop_chunks(newer_than => interval '10 years', table_name => 'disttable_drop_chunks');
 SELECT * FROM disttable_drop_chunks;
 
-CREATE TABLE "weird nAme\\#^."(time bigint, device int CHECK (device > 0), color int, PRIMARY KEY (time,device));
-SELECT * FROM create_distributed_hypertable('"weird nAme\\#^."', 'time', 'device', 3, chunk_time_interval => 100, replication_factor => 2);
+CREATE TABLE "weird nAme\\#^"(time bigint, device int CHECK (device > 0), color int, PRIMARY KEY (time,device));
```
Comment (erimatnor) [inline tsl/test/sql/dist_hypertable.sql.in:850]: Why is this period removed here? This seems like an unrelated and unexplained change for this PR.

Comment (mkindahl) [reply tsl/test/sql/dist_hypertable.sql.in:850]: The period is removed because the name need to be a valid REGCLASS identifier, and periods cannot be part of that and will give a syntax error.

```
mats=# CREATE TABLE "foo$#^." ();
CREATE TABLE
mats=# SELECT 'foo$#^.'::regclass;
ERROR:  invalid name syntax
LINE 1: SELECT 'foo$#^.'::regclass;
               ^
mats=# CREATE TABLE "foo$#^" ();
CREATE TABLE
mats=# SELECT 'foo$#^'::regclass;
 regclass 
----------
 "foo$#^"
(1 row)
```

File: src/chunk.c [hunk 7]
```diff
@@ -3393,47 +3381,99 @@ list_return_srf(FunctionCallInfo fcinfo)
...
+ */
+static Hypertable *
+find_hypertable_for_drop_by_relid(Oid relid, bool force)
+{
+	Hypertable *ht;
+	const char *schema_name;
+	const char *table_name;
+	ContinuousAgg *cagg;
+
+	if ((ht = ts_hypertable_get_by_relid(relid)))
```
[the remark this answers is not in the document]
Comment (mkindahl) [reply src/chunk.c:3434]: For this particular case, there is no significant difference, between the two styles, but for this case, I find this alternative

```
	if (schema_name && table_name &&
		(cagg = ts_continuous_agg_find_userview_name(schema_name, table_name)) &&
		(ht = ts_hypertable_get_by_id(cagg->data.mat_hypertable_id)))
	{
		return ht;
	}
```

a lot clearer than:
```
	if (schema_name && table_name) {
		ContinuousAgg *cagg = ts_continuous_agg_find_userview_name(schema_name, table_name);
		if (cagg) {
			Hypertable *ht = ts_hypertable_get_by_id(cagg->data.mat_hypertable_id);
			if (ht) 
				return ht;
		}
	}
```

And by extension of the suggestion above, I should fix this case as well.

File: src/chunk.c [hunk 8]
```diff
@@ -3393,47 +3381,99 @@ list_return_srf(FunctionCallInfo fcinfo)
...
+ * Find either the hypertable or the materialized hypertable, if the relid is
+ * a continuous aggregate, for the relid.
+ *
+ * Will error if relid is not a hypertable or view (or cannot be found).
+ *
+ * If `force` is false, it will also error if the hypertable found is a
+ * materialized hypertable.
+ */
+static Hypertable *
+find_hypertable_for_drop_by_relid(Oid relid, bool force)
```
[the remark this answers is not in the document]
Comment (mkindahl) [reply src/chunk.c]: This is similar to the parameter for other similar functions and it is also exposed in a PG function, so I think `force` is the best option.

**Comment by k-rus** [comment]
> An option `force` was also added to allow executing `drop_chunks` on
> materialized hypertables on data nodes.

@mkindahl What do you mean by `materialized hypertable`? I don't remember that we used the term ever. Can you use something more familiar?

File: tsl/test/sql/dist_hypertable.sql.in (hunk 6 above)
Comment (k-rus) [reply tsl/test/sql/dist_hypertable.sql.in:850]: It is necessary to use double quotes around the weird name, so it works. In the above example, it will be:
```sql
SELECT '"foo$#^."'::regclass;
```

Comment (k-rus) [reply tsl/test/sql/dist_hypertable.sql.in:850]:
```sql
postgres=# create table "foo$#^." ();
CREATE TABLE
postgres=# select '"foo$#^."'::regclass;
 regclass
-----------
 "foo$#^."
(1 row)

```

Comment (erimatnor) [reply tsl/test/sql/dist_hypertable.sql.in:850]: @mkindahl I am not following. In the existing test, the table is successfully created using the name with the period in it (no syntax error), so it seems it is a valid name. If your issue is casting to a regclass, then I guess what @k-rus suggest is the proper way to do it, not changing the name.

File: src/chunk.c [hunk 9]
```diff
@@ -3376,47 +3367,107 @@ list_return_srf(FunctionCallInfo fcinfo)
...
+			elog(ERROR, "cannot drop chunks on a continuous aggregate materialization table");
+	}
+	else
+	{
+		/* Using the name to look up the continuous aggregate, but we should
+		 * refactor the code to use the relid directly. */
+		const char *const schema_name = get_namespace_name(get_rel_namespace(relid));
+		const char *const table_name = get_rel_name(relid);
+		ContinuousAgg *const cagg = ts_continuous_agg_find_userview_name(schema_name, table_name);
+		Assert(schema_name && table_name);
```
Comment (erimatnor) [inline src/chunk.c:3429]: I am still not sure this is correct. For instance, if you do `drop_chunks(3533, ...)` it will crash (I think). 3533 is neither `InvalidOid` nor a real relation.

You need one of two things I guess:

1. A previous check that relid is a real relation, in which case you can Assert names here, or
2. check for NULL names and throw an error ("invalid relation") if it does not exist

File: tsl/test/sql/dist_hypertable.sql.in (hunk 6 above)
Comment (mkindahl) [reply tsl/test/sql/dist_hypertable.sql.in]: It just replaces one unrelated change with another. The reason we had this strange table name in the first place is because we parsed it ourselves and we wanted to check that it worked. Now we leave that the `regclassin` so the test is not really useful and the actual name is not relevant since it is "whatever reclassin handles". I just fixed the immediate problem in the fastest possible way since I did not consider this test nor what specific name used particularly important.

Anyway, I put the dot back in the name.

**Comment by mkindahl** [comment]
> > An option `force` was also added to allow executing `drop_chunks` on
> > materialized hypertables on data nodes.
> 
> @mkindahl What do you mean by `materialized hypertable`? I don't remember that we used the term ever. Can you use something more familiar?

```
mats@mats-XPS-13-9380:~/repos/timescaledb-chunks$ find src tsl -name '*.c' -exec grep -i "materialized hypertable" {} +
tsl/src/bgw_policy/drop_chunks_api.c:                                    errmsg("cannot add drop chunks policy to materialized hypertable \"%s\" ",
```

Do you have a suggestion for another name?

File: src/chunk.c (hunk 9 above)
Comment (mkindahl) [reply src/chunk.c:3429]: This is checked inside `ts_hypertable_cache_get_entry` but I used the wrong flags so it did not generate an error (fixed that).

Added these tests to make it clear that it works:
```
-- Should error because NULL was used for the table name.
SELECT drop_chunks(NULL, older_than => 3);
ERROR:  invalid hypertable or continuous aggregate
-- should error because there is no relation with that OID.
SELECT drop_chunks(3533, older_than => 3);
ERROR:  OID 3533 does not refer to a table
```

**Comment by k-rus** [comment]
> > > An option `force` was also added to allow executing `drop_chunks` on
> > > materialized hypertables on data nodes.
> > 
> > 
> > @mkindahl What do you mean by `materialized hypertable`? I don't remember that we used the term ever. Can you use something more familiar?
> 
> ```
> mats@mats-XPS-13-9380:~/repos/timescaledb-chunks$ find src tsl -name '*.c' -exec grep -i "materialized hypertable" {} +
> tsl/src/bgw_policy/drop_chunks_api.c:                                    errmsg("cannot add drop chunks policy to materialized hypertable \"%s\" ",
> ```
> 
> Do you have a suggestion for another name?

@mkindahl 
After reading the code, I guess it means the hypertable, which materializes a cagg. But I am really confused with the sentence, which speaks about materialized hyeprtable on data node and option force. Can you elaborate on the commit message? I still don't understand it.

**Comment by mkindahl** [comment]
> > > > An option `force` was also added to allow executing `drop_chunks` on
> > > > materialized hypertables on data nodes.
> > > 
> > > 
> > > @mkindahl What do you mean by `materialized hypertable`? I don't remember that we used the term ever. Can you use something more familiar?
> > 
> > 
> > ```
> > mats@mats-XPS-13-9380:~/repos/timescaledb-chunks$ find src tsl -name '*.c' -exec grep -i "materialized hypertable" {} +
> > tsl/src/bgw_policy/drop_chunks_api.c:                                    errmsg("cannot add drop chunks policy to materialized hypertable \"%s\" ",
> > ```
> > 
> > 
> > Do you have a suggestion for another name?
> 
> @mkindahl
> After reading the code, I guess it means the hypertable, which materializes a cagg. But I am really confused with the sentence, which speaks about materialized hyeprtable on data node and option force. Can you elaborate on the commit message? I still don't understand it.

I elaborated on the commit message some please have a look.

**Comment by k-rus** [comment]
> I elaborated on the commit message some please have a look.

You haven't copied new commit message to the OP, right? Can you copy, so it will be easier to discuss if necessary?

**Comment by mkindahl** [comment]
> > I elaborated on the commit message some please have a look.
> 
> You haven't copied new commit message to the OP, right? Can you copy, so it will be easier to discuss if necessary?

Pasted the log message into the comment above.

File: src/chunk.c [hunk 10]
```diff
@@ -3376,47 +3367,107 @@ list_return_srf(FunctionCallInfo fcinfo)
...
```
Comment (k-rus) [inline src/chunk.c:3429]: Variable declaration section is usually separated by new line from assertions and body:
```suggestion
		ContinuousAgg *const cagg = ts_continuous_agg_find_userview_name(schema_name, table_name);
		
		Assert(schema_name && table_name);
```

File: src/chunk.c [hunk 11]
```diff
@@ -3376,47 +3367,107 @@ list_return_srf(FunctionCallInfo fcinfo)
...
+		if (!allows_materialization && (status & HypertableIsMaterialization))
+			elog(ERROR, "cannot drop chunks on a continuous aggregate materialization table");
+	}
+	else
+	{
+		/* Using the name to look up the continuous aggregate, but we should
+		 * refactor the code to use the relid directly. */
+		const char *const schema_name = get_namespace_name(get_rel_namespace(relid));
+		const char *const table_name = get_rel_name(relid);
+		ContinuousAgg *const cagg = ts_continuous_agg_find_userview_name(schema_name, table_name);
```
Comment (k-rus) [inline src/chunk.c:3428]: If `allows_materialization` is `false`, should it be an error to drop chunks on `cagg`?

File: sql/ddl_api.sql [hunk 12]
```diff
@@ -89,16 +89,14 @@ CREATE OR REPLACE FUNCTION  set_number_partitions(
...
 CREATE OR REPLACE FUNCTION drop_chunks(
+    table_name  REGCLASS,
     older_than "any" = NULL,
-    table_name  NAME = NULL,
-    schema_name NAME = NULL,
     newer_than "any" = NULL,
     verbose BOOLEAN = FALSE,
-    cascade_to_materializations BOOLEAN = NULL
+    cascade_to_materializations BOOLEAN = NULL,
+    allows_materialization BOOLEAN = FALSE
```
[the remark this answers is not in the document]
Comment (mkindahl) [reply sql/ddl_api.sql:99]: It is for internal usage, but exposed because this is how we communicate with the data node. I called it `force` originally, but was asked to change it. If you do not have a concrete suggestion for another name, I will leave it as it is.

File: src/chunk.c [hunk 13]
```diff
@@ -133,6 +133,7 @@ static void init_scan_by_qualified_table_name(ScanIterator *iterator, const char
 #define ASSERT_IS_VALID_CHUNK(chunk) Assert(IS_VALID_CHUNK(chunk))
 
 #define ASSERT_IS_NULL_OR_VALID_CHUNK(chunk) Assert(chunk == NULL || IS_VALID_CHUNK(chunk))
+static void lock_referenced_tables(Oid table_relid);
```
[the remark this answers is not in the document]
Comment (mkindahl) [reply src/chunk.c:130]: It is trivial to ignore this line when reviewing and go to the actual function. I will remove it if I have to do other updates, otherwise I'll leave it.

File: src/chunk.c [hunk 14]
```diff
@@ -3214,39 +3215,30 @@ ts_chunk_drop_process_materialization(Oid hypertable_relid,
...
-	int32 hypertable_id = ts_hypertable_relid_to_id(table_relid);
+	const int32 hypertable_id = ht->fd.id;
 	bool has_continuous_aggs;
 	List *data_nodes = NIL;
 
-	Assert(OidIsValid(table_relid));
-
-	ts_hypertable_permissions_check(table_relid, GetUserId());
+	ts_hypertable_permissions_check(ht->main_table_relid, GetUserId());
+	lock_referenced_tables(ht->main_table_relid);
```
[the remark this answers is not in the document]
Comment (mkindahl) [reply src/chunk.c:3234]: I have been asked to move the comment around already several times, I am not moving it again. It is trivial to go to the function and read the comment there.

File: src/chunk.c [hunk 15]
```diff
@@ -3376,47 +3367,107 @@ list_return_srf(FunctionCallInfo fcinfo)
...
 
-	/* Make sure the schema is set in case it was not given on the command line */
-	FC_ARG(fcinfo, 2) = NameGetDatum(schema_name);
-	FC_NULL(fcinfo, 2) = false;
+	/* this list is from the relcache and can disappear with a cache flush, so
+	 * no further catalog access till we save the fk relids */
+	cachedfkeys = RelationGetFKeyList(table_rel);
+	foreach (lf, cachedfkeys)
+	{
+		ForeignKeyCacheInfo *cachedfk = (ForeignKeyCacheInfo *) lfirst(lf);
```
[the remark this answers is not in the document]
Comment (mkindahl) [reply src/chunk.c]: Agree.

File: src/chunk.c [hunk 16]
```diff
@@ -3376,47 +3367,107 @@ list_return_srf(FunctionCallInfo fcinfo)
...
+
+		/* conrelid should always be that of the table we're considering */
+		Assert(cachedfk->conrelid == RelationGetRelid(table_rel));
+		fk_relids = lappend_oid(fk_relids, cachedfk->confrelid);
+	}
+	table_close(table_rel, AccessShareLock);
 
-	ts_cm_functions->func_call_on_data_nodes(fcinfo, data_node_oids);
+	/* We have a FK between hypertable H and PAR. Hypertable H has number of
```
[the remark this answers is not in the document]
Comment (mkindahl) [reply src/chunk.c:3390]: I have been asked to move these comments around several times already, I'm not moving them again. It is trivial to go to the function and read the comment there.

File: src/chunk.c [hunk 17]
```diff
@@ -3376,47 +3367,107 @@ list_return_srf(FunctionCallInfo fcinfo)
...
+ * Find either the hypertable or the materialized hypertable, if the relid is
+ * a continuous aggregate, for the relid.
+ *
+ * Will error if relid is not a hypertable or view (or cannot be found).
+ *
+ * If `allows_materialization` is false, it will also error if the hypertable
+ * found is a materialized hypertable.
+ */
+static Hypertable *
+find_hypertable_for_drop_by_relid(Cache *hcache, Oid relid, bool allows_materialization)
```
[the remark this answers is not in the document]
Comment (mkindahl) [reply src/chunk.c:3413]: I've been asked to change the name several times already, I'm not changing it again.  It is clear enough.

File: src/chunk.c (hunk 11 above)
Comment (mkindahl) [reply src/chunk.c:3428]: No, this is for internal usage but we have to expose it because that is how we communicate with the data nodes.

File: src/chunk.c [hunk 18]
```diff
@@ -3376,47 +3367,107 @@ list_return_srf(FunctionCallInfo fcinfo)
...
+	List *dc_names = NIL;
+	Oid relid = PG_ARGISNULL(0) ? InvalidOid : PG_GETARG_OID(0);
 	Datum older_than_datum, newer_than_datum;
 	Oid older_than_type, newer_than_type;
 	bool verbose;
 	CascadeToMaterializationOption cascades_to_materializations;
 	int elevel;
-	bool user_supplied_table_name = true;
 	List *data_node_oids = NIL;
+	const bool allows_materialization = PG_ARGISNULL(5) ? false : PG_GETARG_BOOL(5);
```
[the remark this answers is not in the document]
Comment (mkindahl) [reply src/chunk.c:3470]: I've been asked to change the name already. If you do not have a concrete suggestion, I will leave it as it is.

File: src/chunk.c [hunk 19]
```diff
@@ -3376,47 +3367,107 @@ list_return_srf(FunctionCallInfo fcinfo)
...
+}
+
+/*
+ * Find either the hypertable or the materialized hypertable, if the relid is
+ * a continuous aggregate, for the relid.
+ *
+ * Will error if relid is not a hypertable or view (or cannot be found).
+ *
+ * If `allows_materialization` is false, it will also error if the hypertable
+ * found is a materialized hypertable.
```
[the remark this answers is not in the document]
Comment (mkindahl) [reply src/chunk.c:3410]: It means it has to be true to be able to drop chunks directly from the materialization table.

**Comment by mkindahl** [comment]
> I don't see tests of `allows_materialization` on data nodes with caggs. Can you add tests if there are no such?

It is used internally and tests will start failing if the name is changes, so I see no reason to add explicit tests for it.

File: src/chunk.c (hunk 13 above)
[the remark this answers is not in the document]
Comment (k-rus) [reply src/chunk.c:130]: Moving the implementation before function, which uses it, will simplifies reviewing, while ignoring this line doesn't help.

If you hesitant to move the implementation, then I insist to move this declaration line to the same area as other declarations. I don't see why it is good to place it separately and not together with other declarations.
This is common practice to group declarations together. See already existing group at https://github.com/timescale/timescaledb/blob/master/src/chunk.c#L108-L117

File: sql/ddl_api.sql (hunk 12 above)
[the remark this answers is not in the document]
Comment (k-rus) [reply sql/ddl_api.sql:99]: I don't see that this parameter is set internally. Can you point to the place? My current understanding is that users need to set it.

The parameter is exposed publicly, so the name should be meaningful to users. I don't understand why it is needed and its semantics, thus it is difficult for me to propose the name.

File: src/chunk.c (hunk 14 above)
[the remark this answers is not in the document]
Comment (k-rus) [reply src/chunk.c:3234]: It is not trivial, since it is deep inside implementation. Furthermore, the function might be shared with other code paths in future, where the comment will not be valid anymore. 
I do write comment, because it is important, and not just for fun. So I don't understand your ignorance.

It is important to describe here why lock is needed. It is not necessary to have long description.

File: sql/ddl_api.sql (hunk 12 above)
[the remark this answers is not in the document]
Comment (mkindahl) [reply sql/ddl_api.sql:99]: Inside `src/bgw_policy/job.c`:
```
	/* Invoke drop chunks via fmgr so that the call can be deparsed and sent
	 * also to remote data nodes. */
	num_dropped = chunk_invoke_drop_chunks(table_relid,
										   older_than,
										   older_than_type,
										   args->cascade_to_materializations,
										   true);
```

File: src/chunk.c (hunk 16 above)
[the remark this answers is not in the document]
Comment (k-rus) [reply src/chunk.c:3390]: This comment is not meaningful at this place, since it describes semantics of the caller.

Since you complain that you were asked to move the comment several times, this indicates that there is some issue with the code and comment, but nobody can grasp it. Remember that you were providing similar hopeless reviews and now you on other side where you need patience, so the code results to be understandable and maintainable,.

File: src/chunk.c (hunk 19 above)
[the remark this answers is not in the document]
Comment (k-rus) [reply src/chunk.c:3410]: > It means it has to be true to be able to drop chunks directly from the materialization table.

What is the purpose? Is it to prevent users from dropping chunks from the materialised hypertable directly? Or is there other cases when it can happen and this prevention is needed?

File: src/chunk.c (hunk 17 above)
[the remark this answers is not in the document]
Comment (k-rus) [reply src/chunk.c:3413]: If it was clear, I would not comment. I spend good amount of time before doing this request.

File: scripts/clang_format_wrapper.sh [hunk 20]
```diff
@@ -67,7 +67,7 @@ echo "formatting"
 
 cd ${TEMP_DIR}
 
-clang-format ${CLANG_FORMAT_FLAGS} ${FILE_NAMES}
+clang-format-8 ${CLANG_FORMAT_FLAGS} ${FILE_NAMES}
```
[the remark this answers is not in the document]
Comment (erimatnor) [reply scripts/clang_format_wrapper.sh:70]: still there.

File: sql/ddl_api.sql (hunk 12 above)
[the remark this answers is not in the document]
Comment (erimatnor) [reply sql/ddl_api.sql:99]: I am also confused by this new parameter. 

FWIW, my comment on `force` was for an internal function to find a hypertable. Not this public API.

I don't think this parameter was ever discussed or agreed upon. This PR was supposed to remove the wildcard option for drop_chunks, not introduce new parameters that seem unrelated to the original change.

[the remark this answers is not in the document]
Comment (k-rus) [reply sql/ddl_api.sql:99]: I fully agree with @erimatnor I don't understand why this PR should add new parameter, since it is about requiring table name in drop chunks public SQL function.

File: src/chunk.c (hunk 18 above)
[the remark this answers is not in the document]
Comment (k-rus) [reply src/chunk.c:3470]: >I've been asked to change the name already. If you do not have a concrete suggestion, I will leave it as it is.

I've read the code in this PR and I don't understand why this parameter is needed. If I don't mistake, caggs are not supported for distributed hypertables.
It is difficult to propose the name without understanding the parameter.

File: sql/ddl_api.sql (hunk 12 above)
[the remark this answers is not in the document]
Comment (erimatnor) [reply sql/ddl_api.sql:99]: We don't yet support caggs on distributed hypertables and haven't discussed how it should work. So, I don't think this PR should be concerned with drop_chunks on caggs.

File: scripts/clang_format_wrapper.sh (hunk 20 above)
[the remark this answers is not in the document]
Comment (mkindahl) [reply scripts/clang_format_wrapper.sh]: I have to edit the file each time since the CMake does not work correctly, and it is added each time I do `git commit -a` and forget to stash the file.

File: src/chunk.c (hunk 14 above)
[the remark this answers is not in the document]
Comment (mkindahl) [reply src/chunk.c:3267]: Comment moved.

File: src/chunk.c (hunk 11 above)
Comment (mkindahl) [reply src/chunk.c:3433]: Removed `allow_materialization`.

File: sql/ddl_api.sql (hunk 12 above)
[the remark this answers is not in the document]
Comment (mkindahl) [reply sql/ddl_api.sql]: Removed `allow_materialization`.

File: src/chunk.c (hunk 13 above)
[the remark this answers is not in the document]
Comment (mkindahl) [reply src/chunk.c]: Removed.

File: src/chunk.c (hunk 16 above)
[the remark this answers is not in the document]
Comment (mkindahl) [reply src/chunk.c]: Comment moved.

File: src/chunk.c (hunk 17 above)
[the remark this answers is not in the document]
Comment (mkindahl) [reply src/chunk.c]: Name kept, as agreed.

File: src/chunk.c (hunk 18 above)
[the remark this answers is not in the document]
Comment (mkindahl) [reply src/chunk.c]: Removed `allow_materialization`.

File: src/chunk.c [hunk 21]
```diff
@@ -3210,43 +3210,68 @@ ts_chunk_drop_process_materialization(Oid hypertable_relid,
...
 
-	Assert(OidIsValid(table_relid));
+	ts_hypertable_permissions_check(ht->main_table_relid, GetUserId());
 
-	ts_hypertable_permissions_check(table_relid, GetUserId());
+	/* We have a FK between hypertable H and PAR. Hypertable H has number of
+	 * chunks C1, C2, etc. When we execute "drop table C", PG acquires locks
+	 * on C and PAR. If we have a query as "select * from hypertable", this
+	 * acquires a lock on C and PAR as well. But the order of the locks is not
+	 * the same and results in deadlocks. - github issue 865 We hope to
```
Comment (k-rus) [inline src/chunk.c:3262]: If you don't mind to make the following update, so GitHub introduces hyperlink to the issue:
```suggestion
	 * the same and results in deadlocks. - github issue #865 We hope to
```

**Comment by k-rus** [comment]
@mkindahl I forgot to mention that it might be good to squash the commits into one, since fixes to comments on the refactored lock function didn't go to the lock function refactoring commit, but into the second commit.

**Comment by mkindahl** [comment]
> @mkindahl I forgot to mention that it might be good to squash the commits into one, since fixes to comments on the refactored lock function didn't go to the lock function refactoring commit, but into the second commit.

I will do that before I merge, when I've got approvals.

File: src/chunk.c [hunk 22]
```diff
@@ -3375,48 +3399,80 @@ list_return_srf(FunctionCallInfo fcinfo)
...
 
-	/* Make sure the schema is set in case it was not given on the command line */
-	FC_ARG(fcinfo, 2) = NameGetDatum(schema_name);
-	FC_NULL(fcinfo, 2) = false;
+	ht = ts_hypertable_cache_get_entry(hcache, relid, CACHE_FLAG_MISSING_OK);
+	if (ht)
+	{
+		const ContinuousAggHypertableStatus status = ts_continuous_agg_hypertable_status(ht->fd.id);
+		if (status & HypertableIsMaterialization)
+			elog(ERROR, "cannot drop chunks on a continuous aggregate materialization hypertable");
```
[the remark this answers is not in the document]
Comment (mkindahl) [reply src/chunk.c]: Changed message.

File: test/expected/chunk_utils.out [hunk 23]
```diff
@@ -1077,20 +1050,16 @@ SELECT show_chunks();
...
 ERROR:  can only use "drop_chunks" with an INTERVAL for TIMESTAMP, TIMESTAMPTZ, and DATE types
-SELECT drop_chunks(interval '1 minute', 'drop_chunk_test3');
-ERROR:  can only use "drop_chunks" with an INTERVAL for TIMESTAMP, TIMESTAMPTZ, and DATE types
-SELECT drop_chunks(now()-interval '1 minute', 'drop_chunk_test_ts');
-ERROR:  time constraint arguments of "drop_chunks" should have same type as time column of the hypertable
-SELECT drop_chunks(now()::timestamp-interval '1 minute', 'drop_chunk_test_tstz');
-ERROR:  time constraint arguments of "drop_chunks" should have same type as time column of the hypertable
-SELECT drop_chunks(5, schema_name=>'public', newer_than=>4);
+SELECT drop_chunks'drop_chunk_test_ts', (now()-interval '1 minute');
+ERROR:  type "drop_chunks" does not exist at character 8
```
[the remark this answers is not in the document]
Comment (mkindahl) [reply test/expected/chunk_utils.out]: This seems to have gone away in the latest update. It obvious is a missing parentheses, but I could not find the problem in the latest version of the test file.

File: src/chunk.c [hunk 24]
```diff
@@ -3375,48 +3399,85 @@ list_return_srf(FunctionCallInfo fcinfo)
...
+				 errmsg("OID %u does not refer to a table or view", relid)));
 
-	/* Make sure the schema is set in case it was not given on the command line */
-	FC_ARG(fcinfo, 2) = NameGetDatum(schema_name);
-	FC_NULL(fcinfo, 2) = false;
+	ht = ts_hypertable_cache_get_entry(hcache, relid, CACHE_FLAG_MISSING_OK);
+	if (ht)
+	{
+		const ContinuousAggHypertableStatus status = ts_continuous_agg_hypertable_status(ht->fd.id);
+		if (status & HypertableIsMaterialization)
```
[the remark this answers is not in the document]
Comment (mkindahl) [reply src/chunk.c:3425]: It is not necessary since `ContinuousAggHypertableStatus` is a bitmask:

```
typedef enum ContinuousAggHypertableStatus
{
  HypertableIsNotContinuousAgg = 0,
  HypertableIsMaterialization = 1,
  HypertableIsRawTable = 2,
  HypertableIsMaterializationAndRaw = HypertableIsMaterialization | HypertableIsRawTable,
} ContinuousAggHypertableStatus;
```

File: sql/ddl_api.sql [hunk 25]
```diff
@@ -89,13 +89,11 @@ CREATE OR REPLACE FUNCTION  set_number_partitions(
     dimension_name          NAME = NULL
 ) RETURNS VOID AS '@MODULE_PATHNAME@', 'ts_dimension_set_num_slices' LANGUAGE C VOLATILE;
 
--- Drop chunks older than the given timestamp. If a hypertable name is given,
--- drop only chunks associated with this table. Any of the first three arguments
--- can be NULL meaning "all values".
+-- Drop chunks older than the given timestamp for the specific
+-- hypertable or continuous aggregate.
 CREATE OR REPLACE FUNCTION drop_chunks(
+    hypertable_or_cagg  REGCLASS,
```
[the remark this answers is not in the document]
Comment (mkindahl) [reply sql/ddl_api.sql:95]: I have no strong opinion, but I am reluctant to change this since there has been quite a lot of discussions about choice of names already and changing the name now will open up the discussion again.

File: src/chunk.c (hunk 24 above)
[the remark this answers is not in the document]
Comment (gayyappan) [reply src/chunk.c:3425]: This code changes existing logic. A cagg table cannot be a raw + materialization now. But it is something we might support in the future.. We should continue to explicitly check for this in the code given that it is one of the valid return values for Status.

[the remark this answers is not in the document]
Comment (mkindahl) [reply src/chunk.c:3425]: Not sure what I am missing.

From the typedef, we have

- `HypertableIsNotContinuousAgg` = 0,
- `HypertableIsMaterialization` = 1,
- `HypertableIsRawTable` = 2,
- `HypertableIsMaterializationAndRaw` = `HypertableIsMaterialization | HypertableIsRawTable` = 3

Simplifying the code a little (to just reason about when to print an
error or not and just discussing the `status` variable, we have this
in the old code:

```c
switch (status)
{
	case HypertableIsMaterialization:
		/* error */
		break;
	case HypertableIsMaterializationAndRaw:
		/* error */
 		break;
 	case HypertableIsRawTable:
	    ... /* no error */
 	default:
	    ... /* no error */
}
```

Which gives:

- If `status = 0` then OK
- If `status = 1` then ERROR
- If `status = 2` then OK
- If `status = 3` then ERROR

And this in the new code:

```c
if (status & HypertableIsMaterialization)
	/* error */
```

Which gives:

- If `status = 0` then OK (since `status & 1 == 0`)
- If `status = 1` then ERROR (since `status & 1 != 0`)
- If `status = 2` then OK (since `status & 1 == 0`)
- If `status = 3` then ERROR (since `status & 1 != 0`)

These both produce the same ERROR/OK for the same values of `status`.

What am I missing?
